hnn_core/parallel_backends.py

```diff
@@ -253,43 +254,95 @@ def simulate(self, net):
         cmdargs = shlex.split(self.mpi_cmd_str, posix=use_posix)
 
         pickled_params = codecs.encode(pickle.dumps(net.params),
-                                       "base64").decode()
+                                       "base64")
 
         # set some MPI environment variables
         my_env = os.environ.copy()
-        my_env["OMPI_MCA_btl_base_warn_component_unused"] = '0'
-        # Start the simulation in parallel!
-        proc = Popen(cmdargs, stdin=PIPE, stdout=PIPE, stderr=PIPE, env=my_env,
-                     cwd=os.getcwd(), universal_newlines=True)
+        if 'win' not in sys.platform:
+            my_env["OMPI_MCA_btl_base_warn_component_unused"] = '0'
+
+        if 'darwin' in sys.platform:
+            my_env["PMIX_MCA_gds"] = "^ds12"  # open-mpi/ompi/issues/7516
+            my_env["TMPDIR"] = "/tmp"  # open-mpi/ompi/issues/2956
 
-        # wait until process completes
-        out, err = proc.communicate(pickled_params)
+        # set up pairs of pipes to communicate with subprocess
+        (pipe_stdin_r, pipe_stdin_w) = os.pipe()
+        (pipe_stdout_r, pipe_stdout_w) = os.pipe()
+        (pipe_stderr_r, pipe_stderr_w) = os.pipe()
 
-        # print all messages (including error messages)
-        print(out)
+        # Start the simulation in parallel!
+        proc = Popen(cmdargs, stdin=pipe_stdin_r, stdout=pipe_stdout_w,
+                     stderr=pipe_stderr_w, env=my_env, cwd=os.getcwd(),
+                     universal_newlines=True)
```

Comment:
is the codecov complaining legit? worth adding a test or not?

@rythorpe or @ntolley does any of you want to take a look? Just to increase the bus factor of this part of the code a bit ... so it's not just me maintaining this code in the future.

Reply:
Note the test I added `test_mpi_failure`. It seems codecov is not testing MPIBackend to begin with?

Reply:
humm ... I'm a bit lost, didn't we have this test: https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/tests/test_compare_hnn.py#L88 ?

But codecov shows these lines are not covered: https://codecov.io/gh/jonescompneurolab/hnn-core/pull/185/diff ? What's the reason these lines are not reached by the test above?

Reply:
because it's a separate process, I suspect.

Reply:
I don't quite understand why the code coverage decreased here. Could you unpack this a little @blakecaldwell?

Reply:
@rythorpe @jasmainak The codecov % is not reflective of the code actually tested. Yes, it sounds contradictory to the purpose of codecov, but codecov fails here. To explain, let's start with how codecov actually measures lines of code tested. Instead of calling `python`, you call the `codecov` process, which in turn [forks](https://en.wikipedia.org/wiki/Fork_(system_call)) Python. If all of your code is within that process address space (forked processes share the process address space), then codecov can keep track of it. `Popen()` though spawns a separate system process (not forked), and codecov does not know how to keep track of the lines of Python code that are run by an MPI process.

It looks like codecov has designed [a way to work](https://stackoverflow.com/questions/28297497/python-code-coverage-and-multiprocessing) with Python multiprocessing, which must have initially suffered from the same issue, but that's pure Python. It would be very difficult, but not impossible (there are other code coverage tools), to watch a non-Python proces.

Rather than continuing to dive into this, I suggest we just ignore the codecov report here. :)

Reply:
Got it, that makes sense. Thanks for the explanation @blakecaldwell!